wildApprox failed to remove uninstantiated B

## Compiler version

3.1.2-RC1-bin-20211124-1f48de1-NIGHTLY

## Minimized code

```Scala
val l: List[Option[Int]] = List(None, Some(1), None)

@main def m15 =
  l.collectFirst(Some.unapply.unlift[Option[Int], Int])
```

## Output

<details>
<summary>
Press the triangle for the full stack trace.

```
java.lang.AssertionError: assertion failed: `wildApprox` failed to remove uninstantiated B
```

</summary>

```
[error] java.lang.AssertionError: assertion failed: `wildApprox` failed to remove uninstantiated B
[error] scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error] dotty.tools.dotc.typer.ImplicitRunInfo$collectParts$2$.traverse(Implicits.scala:594)
[error] dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:5956)
[error] dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:5956)
[error] dotty.tools.dotc.core.Types$TypeAccumulator.foldOver(Types.scala:5890)
[error] dotty.tools.dotc.core.Types$TypeTraverser.traverseChildren(Types.scala:5957)
[error] dotty.tools.dotc.typer.ImplicitRunInfo$collectParts$2$.traverse(Implicits.scala:599)
[error] dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:5956)
[error] dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:5956)
[error] dotty.tools.dotc.core.Types$TypeAccumulator.op$proxy24$1(Types.scala:5870)
[error] dotty.tools.dotc.core.Types$TypeAccumulator.foldArgs$3(Types.scala:5870)
[error] dotty.tools.dotc.core.Types$TypeAccumulator.foldOver(Types.scala:5874)
[error] dotty.tools.dotc.core.Types$TypeTraverser.traverseChildren(Types.scala:5957)
[error] dotty.tools.dotc.typer.ImplicitRunInfo$collectParts$2$.traverse(Implicits.scala:599)
[error] dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:5956)
[error] dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:5956)
[error] dotty.tools.dotc.typer.ProtoTypes$PolyProto.fold(ProtoTypes.scala:599)
[error] dotty.tools.dotc.core.Types$TypeAccumulator.foldOver(Types.scala:5919)
[error] dotty.tools.dotc.core.Types$TypeTraverser.traverseChildren(Types.scala:5957)
[error] dotty.tools.dotc.typer.ImplicitRunInfo$collectParts$2$.traverse(Implicits.scala:599)
[error] dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:5956)
[error] dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:5956)
[error] dotty.tools.dotc.typer.ProtoTypes$SelectionProto.fold(ProtoTypes.scala:214)
[error] dotty.tools.dotc.core.Types$TypeAccumulator.foldOver(Types.scala:5919)
[error] dotty.tools.dotc.core.Types$TypeTraverser.traverseChildren(Types.scala:5957)
[error] dotty.tools.dotc.typer.ImplicitRunInfo$collectParts$2$.traverse(Implicits.scala:599)
[error] dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:5956)
[error] dotty.tools.dotc.core.Types$TypeTraverser.apply(Types.scala:5956)
[error] dotty.tools.dotc.typer.ProtoTypes$ViewProto.fold(ProtoTypes.scala:545)
[error] dotty.tools.dotc.core.Types$TypeAccumulator.foldOver(Types.scala:5919)
[error] dotty.tools.dotc.core.Types$TypeTraverser.traverseChildren(Types.scala:5957)
[error] dotty.tools.dotc.typer.ImplicitRunInfo$collectParts$2$.traverse(Implicits.scala:599)
[error] dotty.tools.dotc.typer.ImplicitRunInfo$collectParts$2$.apply(Implicits.scala:604)
[error] dotty.tools.dotc.typer.ImplicitRunInfo.recur$1(Implicits.scala:680)
[error] dotty.tools.dotc.typer.ImplicitRunInfo.computeIScope(Implicits.scala:691)
[error] dotty.tools.dotc.typer.ImplicitRunInfo.implicitScope(Implicits.scala:762)
[error] dotty.tools.dotc.typer.ImplicitRunInfo.implicitScope$(Implicits.scala:548)
[error] dotty.tools.dotc.Run.implicitScope(Run.scala:40)
[error] dotty.tools.dotc.typer.Implicits$ImplicitSearch.implicitScope(Implicits.scala:1485)
[error] dotty.tools.dotc.typer.Implicits$ImplicitSearch.searchImplicit(Implicits.scala:1448)
[error] dotty.tools.dotc.typer.Implicits$ImplicitSearch.searchImplicit(Implicits.scala:1457)
[error] dotty.tools.dotc.typer.Implicits$ImplicitSearch.bestImplicit(Implicits.scala:1482)
[error] dotty.tools.dotc.typer.Implicits.inferImplicit(Implicits.scala:988)
[error] dotty.tools.dotc.typer.Implicits.inferImplicit$(Implicits.scala:776)
[error] dotty.tools.dotc.typer.Typer.inferImplicit(Typer.scala:113)
[error] dotty.tools.dotc.typer.Implicits.inferView(Implicits.scala:814)
[error] dotty.tools.dotc.typer.Implicits.inferView$(Implicits.scala:776)
[error] dotty.tools.dotc.typer.Typer.inferView(Typer.scala:113)
[error] dotty.tools.dotc.typer.Typer.tryExtensionOrConversion(Typer.scala:3220)
[error] dotty.tools.dotc.typer.Typer.tryInsertImplicitOnQualifier$$anonfun$1(Typer.scala:3167)
[error] dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:3056)
[error] dotty.tools.dotc.typer.Typer.tryInsertImplicitOnQualifier(Typer.scala:3173)
[error] dotty.tools.dotc.typer.Typer.tryImplicit$1(Typer.scala:3122)
[error] dotty.tools.dotc.typer.Typer.tryInsertApplyOrImplicit(Typer.scala:3135)
[error] dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:3942)
[error] dotty.tools.dotc.typer.Typer.adapt(Typer.scala:3281)
[error] dotty.tools.dotc.typer.Typer.tryApply$1(Typer.scala:3118)
[error] dotty.tools.dotc.typer.Typer.tryInsertApplyOrImplicit$$anonfun$2(Typer.scala:3136)
[error] dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:3056)
[error] dotty.tools.dotc.typer.Typer.tryInsertApplyOrImplicit(Typer.scala:3145)
[error] dotty.tools.dotc.typer.Typer.adapt1(Typer.scala:3942)
[error] dotty.tools.dotc.typer.Typer.adapt(Typer.scala:3281)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2912)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2916)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3032)
[error] dotty.tools.dotc.typer.Applications.typedTypeApply(Applications.scala:1078)
[error] dotty.tools.dotc.typer.Applications.typedTypeApply$(Applications.scala:317)
[error] dotty.tools.dotc.typer.Typer.typedTypeApply(Typer.scala:113)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2803)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2847)
[error] dotty.tools.dotc.typer.ProtoTypes$FunProto.$anonfun$7(ProtoTypes.scala:431)
[error] dotty.tools.dotc.typer.ProtoTypes$FunProto.cacheTypedArg(ProtoTypes.scala:359)
[error] dotty.tools.dotc.typer.ProtoTypes$FunProto.typedArg(ProtoTypes.scala:432)
[error] dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:853)
[error] dotty.tools.dotc.typer.Applications$ApplyToUntyped.typedArg(Applications.scala:853)
[error] dotty.tools.dotc.typer.Applications$Application.addTyped$1(Applications.scala:544)
[error] dotty.tools.dotc.typer.Applications$Application.matchArgs(Applications.scala:609)
[error] dotty.tools.dotc.typer.Applications$Application.init(Applications.scala:447)
[error] dotty.tools.dotc.typer.Applications$TypedApply.<init>(Applications.scala:735)
[error] dotty.tools.dotc.typer.Applications$ApplyToUntyped.<init>(Applications.scala:852)
[error] dotty.tools.dotc.typer.Applications.ApplyTo(Applications.scala:1051)
[error] dotty.tools.dotc.typer.Applications.ApplyTo$(Applications.scala:317)
[error] dotty.tools.dotc.typer.Typer.ApplyTo(Typer.scala:113)
[error] dotty.tools.dotc.typer.Applications.simpleApply$1(Applications.scala:898)
[error] dotty.tools.dotc.typer.Applications.realApply$1$$anonfun$3(Applications.scala:977)
[error] dotty.tools.dotc.typer.Typer.tryEither(Typer.scala:3056)
[error] dotty.tools.dotc.typer.Applications.realApply$1(Applications.scala:988)
[error] dotty.tools.dotc.typer.Applications.typedApply(Applications.scala:1026)
[error] dotty.tools.dotc.typer.Applications.typedApply$(Applications.scala:317)
[error] dotty.tools.dotc.typer.Typer.typedApply(Typer.scala:113)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2784)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2847)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2912)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2916)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3032)
[error] dotty.tools.dotc.typer.Namer.typedAheadExpr$$anonfun$1(Namer.scala:1430)
[error] dotty.tools.dotc.typer.Namer.typedAhead(Namer.scala:1420)
[error] dotty.tools.dotc.typer.Namer.typedAheadExpr(Namer.scala:1430)
[error] dotty.tools.dotc.typer.Namer.typedAheadRhs$1$$anonfun$1(Namer.scala:1682)
[error] dotty.tools.dotc.typer.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:238)
[error] dotty.tools.dotc.typer.Namer.typedAheadRhs$1(Namer.scala:1682)
[error] dotty.tools.dotc.typer.Namer.rhsType$1(Namer.scala:1690)
[error] dotty.tools.dotc.typer.Namer.cookedRhsType$1(Namer.scala:1708)
[error] dotty.tools.dotc.typer.Namer.lhsType$1(Namer.scala:1709)
[error] dotty.tools.dotc.typer.Namer.inferredResultType(Namer.scala:1720)
[error] dotty.tools.dotc.typer.Namer.inferredType$1(Namer.scala:1469)
[error] dotty.tools.dotc.typer.Namer.valOrDefDefSig(Namer.scala:1476)
[error] dotty.tools.dotc.typer.Namer.defDefSig(Namer.scala:1560)
[error] dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:756)
[error] dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:888)
[error] dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:780)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:168)
[error] dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:188)
[error] dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:190)
[error] dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:370)
[error] dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:2726)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2751)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2846)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2912)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2916)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2938)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2988)
[error] dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:2429)
[error] dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$1(Typer.scala:2772)
[error] dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:2776)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2846)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2912)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2916)
[error] dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:2938)
[error] dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:2988)
[error] dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:2556)
[error] dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:2817)
[error] dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:2847)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2912)
[error] dotty.tools.dotc.typer.Typer.typed(Typer.scala:2916)
[error] dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3032)
[error] dotty.tools.dotc.typer.TyperPhase.liftedTree1$1(TyperPhase.scala:47)
[error] dotty.tools.dotc.typer.TyperPhase.typeCheck$$anonfun$1(TyperPhase.scala:53)
[error] dotty.tools.dotc.core.Phases$Phase.monitor(Phases.scala:411)
[error] dotty.tools.dotc.typer.TyperPhase.typeCheck(TyperPhase.scala:54)
[error] dotty.tools.dotc.typer.TyperPhase.runOn$$anonfun$3(TyperPhase.scala:88)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.immutable.List.foreach(List.scala:333)
[error] dotty.tools.dotc.typer.TyperPhase.runOn(TyperPhase.scala:88)
[error] dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:261)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error] scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error] scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error] dotty.tools.dotc.Run.runPhases$1(Run.scala:272)
[error] dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:280)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[error] dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
[error] dotty.tools.dotc.Run.compileUnits(Run.scala:289)
[error] dotty.tools.dotc.Run.compileSources(Run.scala:222)
[error] dotty.tools.dotc.Run.compile(Run.scala:206)
[error] dotty.tools.dotc.Driver.doCompile(Driver.scala:39)
[error] dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:88)
[error] dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error] sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:192)
[error] scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error] sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:247)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:182)
[error] sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:163)
[error] sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:210)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:175)
[error] sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:173)
[error] sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:459)
[error] sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
[error] sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
[error] sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
[error] sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:414)
[error] sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:501)
[error] sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:401)
[error] sbt.internal.inc.Incremental$.apply(Incremental.scala:167)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:528)
[error] sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:482)
[error] sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
[error] sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
[error] sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
[error] sbt.Defaults$.compileIncrementalTaskImpl(Defaults.scala:2357)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$2(Defaults.scala:2314)
[error] sbt.internal.io.Retry$.apply(Retry.scala:46)
[error] sbt.internal.io.Retry$.apply(Retry.scala:28)
[error] sbt.internal.io.Retry$.apply(Retry.scala:23)
[error] sbt.internal.server.BspCompileTask$.compute(BspCompileTask.scala:31)
[error] sbt.Defaults$.$anonfun$compileIncrementalTask$1(Defaults.scala:2310)
[error] scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] sbt.Execute.work(Execute.scala:291)
[error] sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error] java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error] java.base/java.lang.Thread.run(Thread.java:829)
```

</details>

## Expectation
Show an error because unlift doesn't take type parameters.

--- a/compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala
+++ b/compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala
@@ -582,7 +582,7 @@ object ProtoTypes {
     override def isMatchedBy(tp: Type, keepConstraint: Boolean)(using Context): Boolean =
       canInstantiate(tp) || tp.member(nme.apply).hasAltWith(d => canInstantiate(d.info))
 
-    def derivedPolyProto(targs: List[Tree], resultType: Type): PolyProto =
+    def derivedPolyProto(targs: List[Tree], resType: Type): PolyProto =
       if ((targs eq this.targs) && (resType eq this.resType)) this
       else PolyProto(targs, resType)
 

--- a/tests/neg/i14145.scala
+++ b/tests/neg/i14145.scala
@@ -1,0 +1,4 @@
+val l: List[Option[Int]] = List(None, Some(1), None)
+
+@main def m15 =
+  l.collectFirst(Some.unapply.unlift[Option[Int], Int]) // error

